doc/dev/deduplication.rst

```diff
@@ -0,0 +1,233 @@
+base pool, which contains metadata for data deduplication.
+struct object_manifest_t {
+  enum {
+    TYPE_NONE = 0,
+    TYPE_REDIRECT = 1,
+    TYPE_CHUNKED = 2,
+  };
+  uint8_t type;  // redirect, chunked, ...
+  hobject_t redirect_target;
+  std::map<uint64_t, chunk_info_t> chunk_map;
+  ...
+}
+
+Chunk Object: 
+Chunk objects are stored in chunk pool. Chunk object contains chunk data 
+and its reference count information.
+
+
+Although chunk objects and manifest objects have a different purpose 
+from existing objects, they can be handled the same way as 
+original objects. Therefore, to support existing feature such as replication,
+no additional operations for dedup are needed.
+
+Usage
+=====
+
+To set up deduplication pools, you must have two pools. One will act as the 
+base pool and the other will act as the chunk pool. Base pool need to be
+configured with fingerprint_algorithm option as follows.
+ceph osd pool set $BASE_POOL fingerprint_algorithm sha1|sha256|sha512 
+--yes-i-really-mean-it
+
+1. Create objects
+  rados -p base_pool put foo ./foo
+  rados -p chunk_pool put foo-chunk ./foo-chunk
+
+2. Make a manifest object
+  rados -p base_pool set-chunk foo $START_OFFSET $END_OFFSET --target-pool 
+  chunk_pool foo-chunk $START_OFFSET --with-reference
+
+
+Interface
+=========
+
+* set-redirect 
+  set redirection between a base_object in the base_pool and a target_object 
+  in the target_pool.
+  A redirected object will forward all operations from the client to the 
+  target_object.
+  
+  rados -p base_pool set-redirect <base_object> --target-pool <target_pool> 
+   <target_object>
+
+* set-chunk
+  set chunk-offset in a source_object to make a link between it and a 
+  target_object.
+  
+  rados -p base_pool set-chunk <source_object> <offset> <length> --target-pool 
+   <caspool> <target_object> <taget-offset> 
+
+* tier-promote
+  promote the object (including chunks).
+
+  rados -p base_pool tier-promote <obj-name> 
+
+* unset-manifest
+  unset manifest optioan form the object that has manifest.
+
+  rados -p base_pool unset-manifest <obj-name>
+
+
+Dedup tool
+==========
+
+Dedup tool has two features: finding optimal chunk offset for dedup chunking 
+and fixing the reference count.
+
+* find optimal chunk offset
+
+  a. fixex chunk 
+  To find out a fixed chunk length, you need to run following command many 
+  times while changing the chunk_size.
+
+  ceph-dedup-tool --op estimate --pool $POOL --chunk-size chunk_size  
+    --chunk-algorithm fixed --fingerprint-algorithm sha1|sha256|sha512
+
+  b. rabin chunk
+    Rabin-karp algorithm
+    As you know, Rabin-karp algorithm is string-searching algorithm based
+    on rolling-hash. But rolling-hash is not enough to do deduplication because 
+    we don't know the chunk boundary. So, we need content-based slicing using 
+    a rolling hash for content-defined chunking.
+    Current implementation uses the simplest approach (calculate the dynamic chunks
```

Comment:
The current

Reply:
needs re-phrase; maybe "approach:  look for chunk boundaries by inspecting the rolling hash for patterns (like the)..."